pages/intelligence.md

```diff
@@ -0,0 +1,137 @@
+# Code intelligence
+
+We'd like to replace elixir_sense with our own implmentation for the following bits
+of code intelligence.
+
+ * Go to definition
+ * Find references
+ * Refactoring
+
+In order to have this level of intelligence of elixir code, we need to
+search both beam files and source code to discover relevant parts and
+locations of elixir and erlang code, index them an make them
+searchable. This is a large task that will not be accomplished in one
+commit, and this document will help to lay out the scope of the work,
+as well as collect thoughts and drive the implementation.
+
+Code intelligence will be oragnized into the following discrete areas:
+
+* **Discovery** - Finding points of interest in the project's codebase, dependencies and elixir core modules
+* **Caching** - Writing the discovered items to disk and ensuring the index is up to date with the state of the project at *all times*. The goal here is to never have to delete the `.lexical` directory.
+* **Indexing** - Storing the discovered artifacts in a local in-memory index that enable fast searching.
+* **Searching** ideally using something like the [sublime text's fuzzy search](https://www.forrestthewoods.com/blog/reverse_engineering_sublime_texts_fuzzy_match/), but we definitely need some type of fuzzy search.
+* **Recognition** For several of the above examples (go to definition, find references), we need to be able to recognize what our cursor is hovering over. Is it a module? An atom? a variable? etc.
```

Comment:
declaration or a reference for example :)

Reply:
Definitely useful, but I wonder if https://hexdocs.pm/elixir/Code.Fragment.html#surround_context/3 already covers enough of what we need?

Reply:
Not really, for something like: `foo`, there is no "surround context" to go off of.

Reply:
(i really, really want to use built-in functionality by the way)